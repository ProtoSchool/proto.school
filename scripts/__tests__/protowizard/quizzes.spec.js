const api = require('../../../src/api')
const setup = require('../../../jest/helpers/setup')
const fixtures = require('../../../jest/helpers/fixtures')
const asserts = require('../helpers/asserts')
const runners = require('../helpers/runners')

describe('protowizard', () => {
  let lastTutorialId

  beforeAll(() => {
    lastTutorialId = api.tutorials.list.getLatest().id
  })

  afterEach(() => {
    setup.restoreData(lastTutorialId)
  })

  describe('4. create quiz', () => {
    test('4.1 should add quiz to pristine lesson', async () => {
      const tutorial = fixtures.createTutorial({
        lessons: 1,
        lessonOverride: {
          type: 'multiple-choice'
        }
      })

      await runners.protowizard([
        { type: 'quiz' },
        { confirm: true }, // lesson already exists
        { latestTutorial: true }, // add to latest (test) tutorial
        { lesson: tutorial.lessons[0] }, // select first (test) lesson
        { confirm: true }, // overwrite pristine quiz
        { question: 'Is this my question?', // question
          correctAnswer: 'Right answer', // correct answer
          correctFeedback: 'That is soooo right!' }, // correct feedback
        { incorrectAnswer: 'first wrong answer', // incorrect answer
          incorrectFeedback: 'Try again.' }, // incorrect feedback
        { confirm: true }, // add another incorrect answer
        { incorrectAnswer: 'second wrong answer', // incorrect answer
          incorrectFeedback: 'Nope!' }, // incorrect feedba
        { confirm: false }, // don't add another answer
        { confirm: false } // don't add another quiz
      ])

      const result = api.tutorials.getByUrl(tutorial.url)

      expect(result.lessons).toHaveLength(1)
      console.log(result.lessons[0])
      // TODO: create a result.lessons.choices array that can be referenced elsewhere
      // would then be able to test by checking that each answer is present (in any order)
      // OR sorting and then comparing directly

      // TODO: Add assertions - HAS 1 NEW QUIZ WITH GIVEN QUESTIONS
      // asserts.assertNewQuiz({
      //   context: { lastTutorialId },
      //   expected,
      //   result
      // })
    })

    test('4.2 should add quiz to existing lesson', async () => {
      const tutorial = fixtures.createTutorial({
        lessons: 1,
        lessonOverride: {
          type: 'multiple-choice'
        }
      })

      await runners.protowizard([
        // add quiz just to overwrite later
        { type: 'quiz' },
        { confirm: true }, // lesson already exists
        { latestTutorial: true }, // add to latest (test) tutorial
        { lesson: tutorial.lessons[0] }, // select first (test) lesson
        { confirm: true }, // overwrite pristine quiz
        { question: 'Question to overwrite', // question
          correctAnswer: 'Correct answer to overwrite', // correct answer
          correctFeedback: 'Correct feedback to overwrite' }, // correct feedback
        { incorrectAnswer: 'Incorrect answer to overwrite', // incorrect answer
          incorrectFeedback: 'Incorrect feedback to overwrite' }, // incorrect feedback
        { confirm: false }, // don't add another answer
        { confirm: true }, // add another quiz
        // add quiz and overwrite the one above
        { confirm: true }, // lesson already exists
        { latestTutorial: true }, // add to latest (test) tutorial
        { lesson: tutorial.lessons[0] }, // select first (test) lesson
        { confirm: true }, // overwrite non-pristine quiz
        { question: 'Is this my question?', // question
          correctAnswer: 'Right answer', // correct answer
          correctFeedback: 'That is soooo right!' }, // correct feedback
        { incorrectAnswer: 'first wrong answer', // incorrect answer
          incorrectFeedback: 'Try again.' }, // incorrect feedback
        { confirm: true }, // add another incorrect answer
        { incorrectAnswer: 'second wrong answer', // incorrect answer
          incorrectFeedback: 'Nope!' }, // incorrect feedback
        { confirm: false }, // don't add another answer
        { confirm: false } // don't add another quiz
      ])

      const result = api.tutorials.getByUrl(tutorial.url)

      expect(result.lessons).toHaveLength(1)
      console.log(result.lessons[0])
      // TODO: create a result.lessons.choices array that can be referenced elsewhere
      // would then be able to test by checking that each answer is present (in any order)
      // OR sorting and then comparing directly

      // TODO: Add assertions - HAS 1 NEW QUIZ WITH GIVEN QUESTIONS
      // asserts.assertNewQuiz({
      //   context: { lastTutorialId },
      //   expected,
      //   result
      // })
    })

    test('4.3 should cancel overwrite of non-pristine lesson', async () => {
      const tutorial = fixtures.createTutorial({
        lessons: 1,
        lessonOverride: {
          type: 'multiple-choice'
        }
      })

      await runners.protowizard([
        { type: 'quiz' },
        { confirm: true }, // lesson already exists
        { latestTutorial: true }, // add to latest (test) tutorial
        { lesson: tutorial.lessons[0] }, // select first (test) lesson
        { confirm: true }, // overwrite pristine quiz
        { question: 'Is this my question?', // question
          correctAnswer: 'Right answer', // correct answer
          correctFeedback: 'That is soooo right!' }, // correct feedback
        { incorrectAnswer: 'first wrong answer', // incorrect answer
          incorrectFeedback: 'Try again.' }, // incorrect feedback
        { confirm: true }, // add another incorrect answer
        { incorrectAnswer: 'second wrong answer', // incorrect answer
          incorrectFeedback: 'Nope!' }, // incorrect feedback
        { confirm: false }, // don't add another answer
        { confirm: true }, //  add another quiz
        // start to make a new quiz and then cancel rather than overwrite
        { confirm: true }, // lesson already exists
        { latestTutorial: true }, // add to latest (test) tutorial
        { lesson: tutorial.lessons[0] }, // select first (test) lesson
        { confirm: false } // cancel overwriting non-pristine quiz
      ])

      const result = api.tutorials.getByUrl(tutorial.url)

      expect(result.lessons).toHaveLength(1)
      console.log(result.lessons[0])
      // TODO: create a result.lessons.choices array that can be referenced elsewhere
      // would then be able to test by checking that each answer is present (in any order)
      // OR sorting and then comparing directly

      // TODO: Add assertions - HAS 1 NEW QUIZ WITH GIVEN QUESTIONS
      // asserts.assertNewQuiz({
      //   context: { lastTutorialId },
      //   expected,
      //   result
      // })
    })

    test('4.4 should cancel overwrite of pristine lesson', async () => {
      const tutorial = fixtures.createTutorial({
        lessons: 1,
        lessonOverride: {
          type: 'multiple-choice'
        }
      })

      await runners.protowizard([
        { type: 'quiz' },
        { confirm: true }, // lesson already exists
        { latestTutorial: true }, // add to latest (test) tutorial
        { lesson: tutorial.lessons[0] }, // select first (test) lesson
        { confirm: false } // cancel overwriting pristine quiz
      ])

      const result = api.tutorials.getByUrl(tutorial.url)

      expect(result.lessons).toHaveLength(1)
      console.log(result.lessons[0])
      // TODO: create a result.lessons.choices array that can be referenced elsewhere
      // would then be able to test by checking that each answer is present (in any order)
      // OR sorting and then comparing directly

      // TODO: Add assertions - SHOULD HAVE 1 NEW LESSON MATCHING #PRISTINE# BOILERPLATE
      // asserts.assertPristineQuiz({
      //   context: { lastTutorialId },
      //   expected,
      //   result
      // })
    })
  })
})
