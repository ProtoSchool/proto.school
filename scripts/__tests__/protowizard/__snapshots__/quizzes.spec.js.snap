// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`protowizard 4. create quiz 4.1 should add quiz to pristine lesson 1`] = `"Let's create a multiple-choice quiz! This will only work if you've already created the tutorial and lesson files."`

exports[`protowizard 4. create quiz 4.1 should add quiz to pristine lesson 2`] = `
"Great! You've chosen to build a quiz for:
 ‣ Tutorial: New Tutorial (8)
 ‣ Lesson: Lesson 1"
`

exports[`protowizard 4. create quiz 4.1 should add quiz to pristine lesson 3`] = `"When we create a quiz using the ProtoWizard, it will overwrite any existing content. If you think you may have previously updated this quiz manually, you may want to double check before proceeding."`

exports[`protowizard 4. create quiz 4.1 should add quiz to pristine lesson 4`] = `"To preview the quiz in its current state, first run \`npm start\` in a separate terminal window or tab, then visit this page in your web browser: http://localhost:3000/new-tut-8/01"`

exports[`protowizard 4. create quiz 4.1 should add quiz to pristine lesson 5`] = `"Now it's time to write the question for your multiple-choice quiz and provide answer choices, with positive or negative feedback for each. Wrong answers, and the feedback associated with them, are a great way to address common misconceptions about the topic. Be sure to make your feedback as helpful as possible, to guide the learner to the right choice. You'll need to create 1 correct answer and 2-3 incorrect answers. (I'll take care of randomizing their order later.)"`

exports[`protowizard 4. create quiz 4.1 should add quiz to pristine lesson 6`] = `"Your question is: \\"Is this my question?\\""`

exports[`protowizard 4. create quiz 4.1 should add quiz to pristine lesson 7`] = `"The correct answer is: \\"Right answer\\""`

exports[`protowizard 4. create quiz 4.1 should add quiz to pristine lesson 8`] = `"The feedback provided will be: \\"That is soooo right!\\""`

exports[`protowizard 4. create quiz 4.1 should add quiz to pristine lesson 9`] = `"Your new wrong answer is: \\"first wrong answer\\""`

exports[`protowizard 4. create quiz 4.1 should add quiz to pristine lesson 10`] = `"The feedback provided will be: \\"Try again.\\""`

exports[`protowizard 4. create quiz 4.1 should add quiz to pristine lesson 11`] = `"You currently have 1 correct answer and 1 wrong answers. (We recommend providing 2-3 incorrect options.)"`

exports[`protowizard 4. create quiz 4.1 should add quiz to pristine lesson 12`] = `"Your new wrong answer is: \\"second wrong answer\\""`

exports[`protowizard 4. create quiz 4.1 should add quiz to pristine lesson 13`] = `"The feedback provided will be: \\"Nope!\\""`

exports[`protowizard 4. create quiz 4.1 should add quiz to pristine lesson 14`] = `"You currently have 1 correct answer and 2 wrong answers. (We recommend providing 2-3 incorrect options.)"`

exports[`protowizard 4. create quiz 4.1 should add quiz to pristine lesson 15`] = `"To preview your quiz, first run \`npm start\` in a separate terminal window or tab, then visit this page in your web browser: http://localhost:3000/new-tut-8/01"`

exports[`protowizard 4. create quiz 4.1 should add quiz to pristine lesson 16`] = `"To make changes to your quiz, you can edit this file directly: src/tutorials/0008-new-tut-8/01.js"`

exports[`protowizard 4. create quiz 4.1 should add quiz to pristine lesson 17`] = `"Okay, no problem. You can run the ProtoWizard later to add quizzes."`

exports[`protowizard 4. create quiz 4.1 should add quiz to pristine lesson 18`] = `"View our detailed guide to developing tutorials at: https://bit.ly/protoschool-developing"`

exports[`protowizard 4. create quiz 4.2 should add quiz to existing lesson 1`] = `"Welcome to the ProtoWizard! We're excited to help you build your ProtoSchool tutorial."`

exports[`protowizard 4. create quiz 4.2 should add quiz to existing lesson 2`] = `"Let's create a multiple-choice quiz! This will only work if you've already created the tutorial and lesson files."`

exports[`protowizard 4. create quiz 4.2 should add quiz to existing lesson 3`] = `
"Great! You've chosen to build a quiz for:
 ‣ Tutorial: New Tutorial (8)
 ‣ Lesson: Lesson 1"
`

exports[`protowizard 4. create quiz 4.2 should add quiz to existing lesson 4`] = `"When we create a quiz using the ProtoWizard, it will overwrite any existing content. If you think you may have previously updated this quiz manually, you may want to double check before proceeding."`

exports[`protowizard 4. create quiz 4.2 should add quiz to existing lesson 5`] = `"To preview the quiz in its current state, first run \`npm start\` in a separate terminal window or tab, then visit this page in your web browser: http://localhost:3000/new-tut-8/01"`

exports[`protowizard 4. create quiz 4.2 should add quiz to existing lesson 6`] = `"Now it's time to write the question for your multiple-choice quiz and provide answer choices, with positive or negative feedback for each. Wrong answers, and the feedback associated with them, are a great way to address common misconceptions about the topic. Be sure to make your feedback as helpful as possible, to guide the learner to the right choice. You'll need to create 1 correct answer and 2-3 incorrect answers. (I'll take care of randomizing their order later.)"`

exports[`protowizard 4. create quiz 4.2 should add quiz to existing lesson 7`] = `"Your question is: \\"Question to overwrite\\""`

exports[`protowizard 4. create quiz 4.2 should add quiz to existing lesson 8`] = `"The correct answer is: \\"Correct answer to overwrite\\""`

exports[`protowizard 4. create quiz 4.2 should add quiz to existing lesson 9`] = `"The feedback provided will be: \\"Correct feedback to overwrite\\""`

exports[`protowizard 4. create quiz 4.2 should add quiz to existing lesson 10`] = `"Your new wrong answer is: \\"Incorrect answer to overwrite\\""`

exports[`protowizard 4. create quiz 4.2 should add quiz to existing lesson 11`] = `"The feedback provided will be: \\"Incorrect feedback to overwrite\\""`

exports[`protowizard 4. create quiz 4.2 should add quiz to existing lesson 12`] = `"You currently have 1 correct answer and 1 wrong answers. (We recommend providing 2-3 incorrect options.)"`

exports[`protowizard 4. create quiz 4.2 should add quiz to existing lesson 13`] = `"To preview your quiz, first run \`npm start\` in a separate terminal window or tab, then visit this page in your web browser: http://localhost:3000/new-tut-8/01"`

exports[`protowizard 4. create quiz 4.2 should add quiz to existing lesson 14`] = `"To make changes to your quiz, you can edit this file directly: src/tutorials/0008-new-tut-8/01.js"`

exports[`protowizard 4. create quiz 4.2 should add quiz to existing lesson 15`] = `"Let's create a multiple-choice quiz! This will only work if you've already created the tutorial and lesson files."`

exports[`protowizard 4. create quiz 4.2 should add quiz to existing lesson 16`] = `
"Great! You've chosen to build a quiz for:
 ‣ Tutorial: New Tutorial (8)
 ‣ Lesson: Lesson 1"
`

exports[`protowizard 4. create quiz 4.3 should cancel overwrite of non-pristine lesson 1`] = `"Welcome to the ProtoWizard! We're excited to help you build your ProtoSchool tutorial."`

exports[`protowizard 4. create quiz 4.3 should cancel overwrite of non-pristine lesson 2`] = `"Let's create a multiple-choice quiz! This will only work if you've already created the tutorial and lesson files."`

exports[`protowizard 4. create quiz 4.3 should cancel overwrite of non-pristine lesson 3`] = `
"Great! You've chosen to build a quiz for:
 ‣ Tutorial: New Tutorial (8)
 ‣ Lesson: Lesson 1"
`

exports[`protowizard 4. create quiz 4.3 should cancel overwrite of non-pristine lesson 4`] = `"When we create a quiz using the ProtoWizard, it will overwrite any existing content. If you think you may have previously updated this quiz manually, you may want to double check before proceeding."`

exports[`protowizard 4. create quiz 4.3 should cancel overwrite of non-pristine lesson 5`] = `"To preview the quiz in its current state, first run \`npm start\` in a separate terminal window or tab, then visit this page in your web browser: http://localhost:3000/new-tut-8/01"`

exports[`protowizard 4. create quiz 4.3 should cancel overwrite of non-pristine lesson 6`] = `"Now it's time to write the question for your multiple-choice quiz and provide answer choices, with positive or negative feedback for each. Wrong answers, and the feedback associated with them, are a great way to address common misconceptions about the topic. Be sure to make your feedback as helpful as possible, to guide the learner to the right choice. You'll need to create 1 correct answer and 2-3 incorrect answers. (I'll take care of randomizing their order later.)"`

exports[`protowizard 4. create quiz 4.3 should cancel overwrite of non-pristine lesson 7`] = `"Your question is: \\"Is this my question?\\""`

exports[`protowizard 4. create quiz 4.3 should cancel overwrite of non-pristine lesson 8`] = `"The correct answer is: \\"Right answer\\""`

exports[`protowizard 4. create quiz 4.3 should cancel overwrite of non-pristine lesson 9`] = `"The feedback provided will be: \\"That is soooo right!\\""`

exports[`protowizard 4. create quiz 4.3 should cancel overwrite of non-pristine lesson 10`] = `"Your new wrong answer is: \\"first wrong answer\\""`

exports[`protowizard 4. create quiz 4.3 should cancel overwrite of non-pristine lesson 11`] = `"The feedback provided will be: \\"Try again.\\""`

exports[`protowizard 4. create quiz 4.3 should cancel overwrite of non-pristine lesson 12`] = `"You currently have 1 correct answer and 1 wrong answers. (We recommend providing 2-3 incorrect options.)"`

exports[`protowizard 4. create quiz 4.3 should cancel overwrite of non-pristine lesson 13`] = `"Your new wrong answer is: \\"second wrong answer\\""`

exports[`protowizard 4. create quiz 4.3 should cancel overwrite of non-pristine lesson 14`] = `"The feedback provided will be: \\"Nope!\\""`

exports[`protowizard 4. create quiz 4.3 should cancel overwrite of non-pristine lesson 15`] = `"You currently have 1 correct answer and 2 wrong answers. (We recommend providing 2-3 incorrect options.)"`

exports[`protowizard 4. create quiz 4.3 should cancel overwrite of non-pristine lesson 16`] = `"To preview your quiz, first run \`npm start\` in a separate terminal window or tab, then visit this page in your web browser: http://localhost:3000/new-tut-8/01"`

exports[`protowizard 4. create quiz 4.3 should cancel overwrite of non-pristine lesson 17`] = `"To make changes to your quiz, you can edit this file directly: src/tutorials/0008-new-tut-8/01.js"`

exports[`protowizard 4. create quiz 4.3 should cancel overwrite of non-pristine lesson 18`] = `"Let's create a multiple-choice quiz! This will only work if you've already created the tutorial and lesson files."`

exports[`protowizard 4. create quiz 4.3 should cancel overwrite of non-pristine lesson 19`] = `
"Great! You've chosen to build a quiz for:
 ‣ Tutorial: New Tutorial (8)
 ‣ Lesson: Lesson 1"
`

exports[`protowizard 4. create quiz 4.4 should cancel overwrite of pristine lesson 1`] = `"Welcome to the ProtoWizard! We're excited to help you build your ProtoSchool tutorial."`

exports[`protowizard 4. create quiz 4.4 should cancel overwrite of pristine lesson 2`] = `"Let's create a multiple-choice quiz! This will only work if you've already created the tutorial and lesson files."`

exports[`protowizard 4. create quiz 4.4 should cancel overwrite of pristine lesson 3`] = `
"Great! You've chosen to build a quiz for:
 ‣ Tutorial: New Tutorial (8)
 ‣ Lesson: Lesson 1"
`

exports[`protowizard 4. create quiz 4.4 should cancel overwrite of pristine lesson 4`] = `"When we create a quiz using the ProtoWizard, it will overwrite any existing content. If you think you may have previously updated this quiz manually, you may want to double check before proceeding."`

exports[`protowizard 4. create quiz 4.4 should cancel overwrite of pristine lesson 5`] = `"To preview the quiz in its current state, first run \`npm start\` in a separate terminal window or tab, then visit this page in your web browser: http://localhost:3000/new-tut-8/01"`

exports[`protowizard 4. create quiz 4.4 should cancel overwrite of pristine lesson 6`] = `"Okay, no problem. You can edit your existing quiz directly in this file: src/tutorials/0008-new-tut-8/01.js"`
